File: blind75.js
//1
const twoSum = (nums, target) => {
  let map = {};
  for (let i = 0; i < nums.length; i++) {
    let complement = target - nums[i];
    if (map[complement] !== undefined) {
      return [map[complement], i];
    } else {
      map[nums[i]] = i;
    }
  }
};

//121
const maxProfit = (prices) => {
  let minPrice = prices[0];
  let maxProfit = 0;

  for (let i = 0; i < prices.length; i++) {
    let val = prices[i] - minPrice;
    if (prices[i] < minPrice) {
      minPrice = prices[i];
    } else if (val > maxProfit) {
      maxProfit = val;
    }
  }

  return maxProfit;
};

//217
const containsDuplicate = (nums) => {
  let map = {};
  for (let i = 0; i < nums.length; i++) {
    if (map[nums[i]] !== undefined) {
      return true;
    } else {
      map[nums[i]] = 1;
    }
  }
  return false;
};

//238
const productExceptSelf = (nums) => {
  let length = nums.length;

  let answer = [];
  answer[0] = 1;

  for (let i = 1; i < length; i++) {
    answer[i] = nums[i - 1] * answer[i - 1];
  }
  let R = 1;
  for (let i = length - 1; i >= 0; i--) {
    answer[i] = answer[i] * R;
    R *= nums[i];
  }

  return answer;
};
// testing windows setup environment

File: blind75.py
# 1
class Solution:
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        num_map = {}

        for i, num in enumerate(nums):
            complement = target - num
            if complement in num_map:
                return [num_map[complement], i]
            num_map[num] = i


# 121
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        minPrice = prices[0]
        maxProfit = 0
        for price in prices:
            val = price - minPrice
            if price < minPrice:
                minPrice = price
            elif val > maxProfit:
                maxProfit = val
        return maxProfit


# 217
class Solution(object):
    def containsDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        val = {}
        for num in nums:
            if num in val:
                return True
            else:
                val[num] = 1
        return False
